### ISSUE TITLE
/admin/elearning-properties: Three LMS Platforms Configured, Same Secret Masking Bug + Incomplete Integration

### ISSUE THREAD

**Issue by lizhsr** [SimisRnD/simis-cms#521]
## Executive Summary

The e-learning settings page has three different Learning Management System (LMS) platforms configured, but:

1. **Same security bug:** All secret fields (LRS Secret, Moodle Token, PERLS Secret) display in plain text instead of being masked
2. **Unclear integration:** It's unknown if these integrations are actually being used (similar pattern to BI feature - configuration-only with no UI integration)
3. **Outdated approach:** Modern LMS platforms are moving toward xAPI and AI features; SimIS supports older patterns

**Pattern Emerging:** Multiple Settings pages have the same secret masking bug. This is a systemic issue, not isolated to one page.

---

## Current State

**Three LMS Platforms Supported:**
1. **LRS/xAPI** (Learning Record Store - Experience API)
2. **Moodle** (open-source LMS)
3. **PERLS** (what is this?)

**Form Fields:**

```
Enable e-learning?: [☑] Yes

Enable LRS xAPI?: [☑] Yes
  LRS URL: [http://...]
  LRS Key: [________________]
  LRS Secret: [________________] ← BUG: Should be masked
  LRS Auth Header: [________________] ← BUG: Should be masked

Enable Moodle?: [☑] Yes
  Moodle URL: [http://...]
  Moodle Token: [________________] ← BUG: Should be masked

Enable PERLS?: [☑] Yes
  PERLS URL: [http://...]
  PERLS Client Id: [________________]
  PERLS Secret: [________________] ← BUG: Should be masked
```

**Issues Found:**

### 1. CRITICAL: Secret Fields Not Masked (Systemic Bug)
All secret properties defined in SecretSitePropertiesCommand but displayed in plain text:
- `elearning.lrs.secret` — marked as secret, displayed as plain text ✗
- `elearning.lrs.authHeader` — marked as secret, displayed as plain text ✗
- `elearning.moodle.token` — marked as secret, displayed as plain text ✗
- `elearning.perls.secret` — marked as secret, displayed as plain text ✗

This indicates a **systemic issue with the secret masking logic in site-properties-editor.jsp**. The masking code exists but isn't working for ANY properties on ANY settings page.

**Root Cause Hypothesis:** Property names in database don't match SecretSitePropertiesCommand list, OR the JSP condition `${secretPropertyNames.contains(siteProperty.name)}` is failing.

### 2. Unclear Integration Status
Like BI feature:
- Backend code exists (ElearningCommand, MoodleApiClientCommand, PERLSApiClientCommand, LRS integration)
- API clients for all three platforms are implemented
- ✗ No way to see if integrations are being used
- ✗ No admin UI to embed e-learning content on pages
- ✗ Unknown if this is active or orphaned code

### 3. Three Platforms, Unclear Use Cases
- **LRS/xAPI:** Modern standard for tracking learning everywhere (mobile, VR, simulations), not just in LMS browser
- **Moodle:** Open-source LMS, popular in education
- **PERLS:** Unknown — no code documentation. Acronym unclear. Not a standard LMS.

Why three? Which should SimIS customers use? Why are they all optional toggles?

### 4. No Documentation
- No help text explaining what each platform is
- No guidance on choosing between them
- No explanation of where to get credentials
- No explanation of which is recommended for what use case

### 5. Outdated Standards (Moodle)
Modern LMS platforms are moving toward xAPI for richer tracking, but Moodle integration appears to use older pattern. Research shows (2026):
- xAPI tracks learning anywhere (mobile, VR, simulations, real-world activities)
- Moodle itself now supports xAPI + AI features (content generation, predictive analytics)
- Canvas and Blackboard dominating higher education
- TalentLMS, Docebo, Cornerstone for corporate training

### 6. No Integration UI
- No way to embed Moodle courses on SimIS pages
- No way to embed PERLS content
- No way to show LRS data/reports
- Like BI: configuration-only with no actual usage capability

---

### LRS/xAPI (Learning Record Store / Experience API)

<cite index="19-1">xAPI (Experience API) decouples learning content from the LMS platform. It tracks learning experiences anywhere, across any device (mobile apps, VR simulations, serious games, offline on-the-job training) by sending JSON statements (Actor-Verb-Object, e.g., "Jane Doe - completed - CPR Training VR Simulation") securely to a Learning Record Store via HTTP POST</cite>

**Use Case for Defense Contractors:** Tracking employee training across:
- Classroom instruction
- Online courses
- Simulations and exercises
- Virtual reality training
- On-the-job performance

All tracked in one centralized LRS (Learning Record Store).

**Implementation Status:** Backend code exists to send xAPI statements, but unclear if it's being used.

### Moodle (Open-Source LMS)

<cite index="24-1">Moodle is a highly customizable platform with a large plugin ecosystem, supporting standards like SCORM and xAPI. Recent 2025/2026 updates added AI capabilities: content generation, summarization, and predictive analytics to identify at-risk learners</cite>

**Use Case:** Organizations that want to self-host an LMS and customize extensively.

**Implementation Status:** MoodleApiClientCommand exists. Unclear if any content is embedded or accessed.

### PERLS (Unknown)

No clear documentation in codebase. PERLSAccessTokenCommand and PERLSApiClientCommand exist, suggesting it's a platform with OAuth2 access tokens. Could be:
- PERLS by Curabitur (defunct?)
- Something specific to SimIS's past customers?
- Typo/old acronym?

**Recommendation:** Clarify what PERLS is or remove it.

---

## Modern CMS/LMS Integration Approaches (2026)

Modern CMSs supporting e-learning are moving toward:

1. **Direct LMS Embedding**
   - Canvas LMS embeds courses directly on pages
   - Moodle iframe embedding of specific activities
   - Content preview in CMS before publishing

2. **xAPI Integration**
   - <cite index="27-1">xAPI records learning activity across places and contexts (not just packaged content like SCORM)</cite>
   - Richer data collection: track mobile learning, VR, simulations, real-world activities
   - Dashboard showing learning analytics from LRS

3. **LTI (Learning Tools Interoperability)**
   - <cite index="27-1">LTI used for in-platform launching and grade exchange, direct APIs for deeper data access</cite>
   - Standard protocol for embedding external learning tools

4. **AI Integration**
   - <cite index="24-1">Moodle 2026: AI content generation, AI summarization, predictive analytics for at-risk learners</cite>
   - Some platforms (edzlms) build AI tutoring on top of Moodle

5. **Mobile & Offline**
   - Moodle mobile app allows offline course access
   - xAPI tracks mobile learning separately from web

**What SimIS CMS Offers:**
- Backend integration with three LMS (partial, configuration-only)
- ✗ No embedded course display on pages
- ✗ No LRS analytics dashboard
- ✗ No AI features
- ✗ No LTI support
- ✗ Moodle app unavailable through SimIS

---

### Strategic

1. **Are these integrations being used?**
   - Search customer sites for Moodle/PERLS/LRS courses embedded in content
   - If not used: remove to reduce complexity
   - If used: prioritize UI integration

2. **What is PERLS?**
   - Is it still supported?
   - Should it be removed or highlighted?

3. **Why three LMS platforms?**
   - Were they added incrementally as customers requested them?
   - Should only one (Moodle) be recommended going forward?

### Technical

4. **Why aren't secrets being masked?**
   - Is the property name in database different from SecretSitePropertiesCommand?
   - Is site-properties-editor.jsp condition failing?
   - This is affecting CAPTCHA, BI, and e-learning pages

5. **Should we support modern LMS features?**
   - xAPI analytics dashboard (show learning data from LRS)?
   - Embedded Moodle course display?
   - LTI support for generic LMS compatibility?

---

### Phase 0: Fix Secret Masking (Systemic Issue - P0)

**THIS IS BLOCKING:** The secret masking bug affects CAPTCHA, BI, and e-learning pages. Multiple vulnerabilities.

1. **Debug why masking isn't working:**
   - Run on dev: verify database property names match SecretSitePropertiesCommand
   - Add logging to SitePropertiesEditorWidget to see if secretPropertyNames is populated
   - Debug JSP condition: `${secretPropertyNames.contains(siteProperty.name)}`

2. **Once fixed:** All three properties pages (CAPTCHA, BI, e-learning) will be secured

### Phase 1: Clarify E-Learning Strategy

**Before redesigning:**

1. **Audit usage:** Are these integrations actually being used in production?
   - Search all customer sites for embedded courses
   - Check activity logs for Moodle/PERLS API calls
   - Check LRS for recent statements (xAPI)

2. **Clarify PERLS:** What is it? Should it stay?

3. **Simplify if needed:** If only Moodle is used, remove LRS and PERLS to reduce complexity

4. **Plan integration UI:** How should courses appear in SimIS?
   - Embedded course list on pages?
   - Moodle course iframe embeds?
   - LRS learning record dashboard?

### Phase 2: Modern LMS Features (Post-Migration)

If e-learning is important:
- xAPI analytics dashboard (show learning completion trends)
- Moodle course embedding on pages
- Student progress tracking per course
- AI-powered recommendations (if using Moodle 4.5+)

---

## Timeline Estimate

**Phase 0 (Fix Masking):** 2-4 hours (affects 3 pages: CAPTCHA, BI, e-learning)

**Phase 1 (Clarify Strategy):** 1-2 weeks
- Audit customer usage
- Clarify PERLS or remove
- Plan integration UI

**Phase 2 (Modern Features):** 2-3 weeks (if proceeding)

---

## Systemic Issue: Secret Masking

This is now the third Settings page with the same bug:
- CAPTCHA: reCAPTCHA secret displayed in plain text ✗
- BI: Superset secret displayed in plain text ✗
- E-learning: LRS/Moodle/PERLS secrets displayed in plain text ✗

The masking code exists in site-properties-editor.jsp (lines 76-86), but it's not working. This suggests:

**Possible Root Causes:**
1. **Property name mismatch:** Database stores `elearning.lrs.secret`, SecretSitePropertiesCommand expects `elearning.lrs.key` or something else
2. **JSP condition failure:** `${secretPropertyNames.contains(siteProperty.name)}` isn't evaluating correctly
3. **Rendering order:** Secrets rendered before masking logic runs
4. **Widget misconfiguration:** SitePropertiesEditorWidget not being used for these pages

**Recommendation:** Fix the root cause to secure all properties pages at once.

---

## Related Issues

- github_issue_captcha_settings_security.md (same masking bug)
- github_issue_bi_settings_audit.md (same masking bug)
- /admin/mail-properties (likely same masking bug)
- /admin/ecommerce-properties (likely same masking bug)

**Recommendation:** After fixing masking, audit ALL settings pages for the same pattern.

---

## References

**Standards & Platforms:**
- <cite index="19-1">xAPI: Decouples learning from LMS, tracks across mobile, VR, simulations via HTTP POST to LRS</cite>
- <cite index="22-1">Modern standards: SCORM (legacy), xAPI (modern), cmi5 (next-gen combining both)</cite>
- <cite index="24-1">Moodle 4.5+ supports AI: content generation, summarization, predictive analytics</cite>
- <cite index="23-1">Popular LMS: Moodle (open-source), Canvas (cloud), Blackboard (enterprise), TalentLMS (corporate)</cite>

**Code:**
- ElearningCommand.java (loads LRS/Moodle/PERLS settings)
- MoodleApiClientCommand.java (Moodle API integration)
- PERLSApiClientCommand.java (PERLS integration)
- PERLSAccessTokenCommand.java (PERLS OAuth)
- SecretSitePropertiesCommand.java (lists elearning secrets)

---

## Key Decision Needed

**Before investing time: Is e-learning a priority?**

This feature appears incomplete (configuration-only, similar to BI). If customers aren't using it, defer until after Azure migration. If customers need it, complete the integration and add modern LMS features.

Current state: Three LMS platforms configured but unclear if any are actually being used.

**Comment by lizhsr** [comment]
## Update on the secret-masking portion of this issue

Already fixed in code -- [PR #120](https://github.com/SimisRnD/simis-cms/pull/120) (merged 2026-07-19) added masking for all 19 secret site properties, including `elearning.lrs.secret`, `elearning.lrs.authHeader`, `elearning.moodle.token`, and `elearning.perls.secret` (all four fields this issue names). Verified directly against the current codebase -- the masking mechanism is generic and correct for every settings page, not page-specific.

The live site currently still shows these unmasked because the deployed release predates this fix; that's a separate, already-tracked deployment gap, not a code defect requiring a new fix.

The rest of this issue (three LMS platforms, unclear usage/integration status, what PERLS even is) is unrelated to the masking bug and remains open -- those are the real open questions here.

**Comment by lizhsr** [comment]
## Update: documentation gap closed; integration-status claim was overstated in one direction and understated in another

**Documentation (claim #4, "no help text"):** Fixed by [PR #657](https://github.com/SimisRnD/simis-cms/pull/657) (merged 2026-07-29), which added a help-text block for each of LRS, Moodle, and PERLS on `site-properties-editor.jsp` (lines ~264-271), including a direct answer to "what is PERLS?" (a DoD/ADL-funded mobile microlearning app, now archived following ADL's 2025 shutdown).

**Integration status (claims #2/#6, "configuration-only, no UI integration"):** This claim doesn't hold uniformly across the three platforms -- checked each separately against current code:

- **Moodle is genuinely wired in, not orphaned.** `RemoteCourseListWidget` (`src/main/java/.../presentation/widgets/elearning/RemoteCourseListWidget.java`, registered as widget `remoteCourseList` in `widget-library.xml:120`) renders a real course list by role (student/teacher/manager/any) with launch links and an "Add a Course" button -- any admin can drop this on a page via the visual editor. Separately, `CalendarAjaxMoodleEvents.addMoodleEvents()` merges live Moodle events into the site's calendar widget (`CalendarAjax.java`). So "no way to embed Moodle courses on SimIS pages" is false.
- **LRS/xAPI is genuinely dead code**, more so than the issue suggests. `ElearningCommand.isLRSEnabled()` (`src/main/java/.../application/elearning/ElearningCommand.java:39`) has zero callers anywhere in the codebase -- nothing gates on it, and there's no code path that actually POSTs xAPI statements to the configured LRS URL. (The unrelated `xapi` package/`XapiStatementCommand` used by `HistoryTask` is a local activity-history feature, unconnected to the `elearning.lrs.*` properties -- don't confuse the two.)
- **PERLS embedding is also dead.** `RemoteCourseListWidget` computes `perlsEnabled` (line 47) but never references it again in the method -- an unused local variable; only the Moodle branches are implemented. The only live caller of `PERLSCourseListCommand` is `DatasetDownloadRemoteFileCommand`, which uses it for the unrelated Datasets remote-file-import feature, not for displaying PERLS content to end users.

So the accurate framing is: two of three platforms (LRS, PERLS) are configuration-only with no real usage, matching the issue's suspicion; Moodle is not, contradicting it.

**What's left is a product decision, not a code fix.** Whether to build out real LRS/PERLS usage (xAPI statement forwarding, PERLS content embedding) or remove them requires knowing if any deployed site actually has them enabled -- that's a customer/production-usage question this session can't answer from source alone. Recommend leaving this open pending that input rather than speculatively building integration for platforms that may have zero live usage.

The secret-masking claim (#1) was already covered in the prior comment on this issue -- no change there, still fixed by PR #120.

**Comment by lizhsr** [comment]
Per your decision: keeping LRS, Moodle, and PERLS all configured for historical/future purposes, but sorting Moodle first since it's the only one with a real, working integration today. #887 does that -- pure data reordering (property_order), no functionality removed or changed.

**Comment by lizhsr** [comment]
Closing -- everything here is resolved:
- Secret masking: fixed by #120 (merged 2026-07-19).
- Documentation / "what is PERLS": fixed by #657 (merged 2026-07-29).
- Integration status: diagnosed (Moodle is genuinely wired in via RemoteCourseListWidget + calendar sync; LRS and PERLS are dead code with no real callers) and resolved by decision -- keep all three configured for historical/future purposes, Moodle sorted first -- shipped in #887 (merged).
